synth_node_bin/src/action/advanced_sn_for_s001.rs

```diff
@@ -0,0 +1,181 @@
+        let mut broadcast_msgs_interval = interval(BROADCAST_INTERVAL_SEC);
+
+        loop {
+            tokio::select! {
+                _ = dbg_info_interval.tick() => {
+                    trace_debug_info(synth_node).await;
+                },
+                _ = broadcast_msgs_interval.tick() => {
+                    let _ = broadcast_periodic_msgs(synth_node);
+                },
+                Ok((src, msg)) = synth_node.try_recv_message() => {
+                    handle_rx_msg(synth_node, src, msg).await?;
+                },
+            }
+        }
+    }
+}
+
+async fn trace_debug_info(synth_node: &SyntheticNode) {
+    let peer_infos = synth_node.connected_peer_infos();
+    let peer_cnt = peer_infos.len();
+
+    let mut log = format!("\nNumber of peers: {peer_cnt}\n");
+
+    // Let's sort by the connection's time value.
+    let mut peer_infos: Vec<_> = peer_infos.iter().collect();
+    peer_infos.sort_by(|a, b| a.1.stats().created().cmp(&b.1.stats().created()));
+
+    for (addr, info) in peer_infos.iter() {
+        let stats = info.stats();
+
+        // Align all possible IP addresses (both v4 and v6) vertically
+        const MAX_IPV6_ADDR_LEN: usize = 37;
+
+        log.push_str(&format!(
+            "{addr:>ident$}   ({side:?}) seconds: {time:?}\n",
+            addr = addr,
+            ident = MAX_IPV6_ADDR_LEN,
+            side = info.side(),
+            time = stats.created().elapsed()
+        ));
+    }
+
+    tracing::info!("{log}");
+}
+
+fn broadcast_periodic_msgs(synth_node: &mut SyntheticNode) -> Result<()> {
+    if let Err(e) = broadcast_ping_msg(synth_node) {
+        tracing::warn!("failed to broadcast Ping messages: {e}");
+    } else if let Err(e) = broadcast_addr_msg(synth_node) {
+        tracing::warn!("failed to broadcast Addr messages: {e}");
+    } else if let Err(e) = broadcast_get_addr_msg(synth_node) {
+        tracing::warn!("failed to broadcast GetAddr messages: {e}");
+    }
+
+    Ok(())
+}
+
+fn broadcast_ping_msg(synth_node: &mut SyntheticNode) -> Result<()> {
+    let msg = Message::Ping(Nonce::default());
+
+    for addr in synth_node.connected_peers() {
+        if synth_node.unicast(addr, msg.clone()).is_err() {
+            tracing::error!("failed to send {msg:?} to {addr}\n");
+            anyhow::bail!("connection closed");
+        }
+    }
+
+    Ok(())
+}
+
+fn broadcast_get_addr_msg(synth_node: &mut SyntheticNode) -> Result<()> {
+    let msg = Message::GetAddr;
+
+    for addr in synth_node.connected_peers() {
+        if synth_node.unicast(addr, msg.clone()).is_err() {
+            tracing::error!("failed to send {msg:?} to {addr}\n");
+            anyhow::bail!("connection closed");
+        }
+    }
+
+    Ok(())
+}
+
+async fn handle_rx_msg(
+    synth_node: &mut SyntheticNode,
+    src: SocketAddr,
+    msg: Message,
+) -> Result<()> {
+    tracing::info!("message received from {src}:\n{msg:?}");
+
+    // We are only handling a GetAddr for now.
```

Comment:
Shouldn't we also handle `ping` messages to ensure other hosts that connection is valid?

Reply:
There are automatic auto-reply filters applied before the messages reach the inbound msg queue in the `try_recv_message` API, the only autoreply filter which we don't use here is for `GetAddr` messages so we can handle those manually here

Check line 38:
```rust
msg_filter: MessageFilter::with_all_auto_reply().with_getaddr_filter(Filter::Disabled),
```

Reply:
Got it  :)